### ISSUE TITLE
Problem with narrowed tensor and torch.save()

### ISSUE THREAD

**Issue by navid-mahmoudian** [pytorch/pytorch#44872]
## 🐛 Bug

I noticed that when torch.save() is used on a narrowed version of a tensor, it saves the whole tensor (original tensor), but when using torch.load() it returns the narrowed tensor. To me, this looks like a bug because it saves something more than what it is intended, and we are not able to recover this redundant info. For that, I wrote a simple example, and you can check the output to see what I mean:

```
import torch
import os


x= torch.randint(0, 100, size=(10000,8), dtype=torch.long)

y1 = x.narrow(dim=0, start=0, length=10)
torch.save({"y": y1},"y1.pth")

y2 = x.narrow(dim=0, start=0, length=10).detach().clone()
torch.save({"y": y2},"y2.pth")


loaded_y1 = torch.load("y1.pth")["y"]
loaded_y2 = torch.load("y2.pth")["y"]

if torch.equal(loaded_y1, loaded_y2):
    print("loaded_y1==loaded_y2")

if torch.equal(loaded_y1, y1):
    print("loaded_y1==y1")

print("size y1.pth (in Bytes)=", os.stat("y1.pth").st_size)
print("size y2.pth (in Bytes)=", os.stat("y2.pth").st_size)
```
This is the output of the above script:

> loaded_y1==loaded_y2
> loaded_y1==y1
> size y1.pth (in Bytes)= 640754
> size y2.pth (in Bytes)= 1394

As you can see, both “y1.pth” and “y2.pth” are equal, but “y1.pth” is much larger because it saves the mother tensor, while we are not able to recover it.

 - PyTorch Version (e.g., 1.0): 1.6.0+cpu
 - OS (e.g., Linux): Linux (Ubuntu)
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: Python 3.8.2
 - CUDA/cuDNN version: no Cuda

**Comment by glaringlee** [comment]
@albanD 
I am not sure whether this is related to extra info stored for autograd. `x` is created with autograd off, not sure whether extra info will be created when doing y1 = x.narrow(). save() stored more info for `y1` which is the result of calling `narrow`, but it saved much less info for `y2`  which is cloned from y1. Any thoughts?

**Comment by albanD** [comment]
Hi,

No this is expected.
The solution if you want to avoid this is just to `.clone()` the Tensor before saving it. Note that the `.detach()` doesn't change anything here! The serialization is **not** differentiable.

The detail here is just that when you do narrow, the underlying storage actually is still full size. And the serialization stores the full Storage.
In your example, you could actually recover all the other values of x (even though I don't recommend you do it) by doing `loaded_y1.as_strided(x.size(), x.stride(), x.storage_offset())`.

I though we added a note in the doc about this, cc @ailzhang do you remember where it is?

**Comment by navid-mahmoudian** [comment]
Hello, 
Thank you very much for your response. From the beginning, I knew that this is because in narrow the underlying storage is still full size. However, **I do not agree that this is what should be expected**. When users save a narrowed tensor, they expect to save only what is needed. This behavior is what happens in NumPy (i.e. the np.save only saves what is needed, not the full size). Here is a simple NumPy example to show you that NumPy only saves the sliced array:

```

import numpy as np
import os

x = np.random.randint(0, 100, size=(10000,8), dtype=np.int64)

y1 = x[0:10]
np.save("y1.npy", y1)

y2 = np.copy(x[0:10])
np.save("y2.npy", y2)

loaded_y1 = np.load("y1.npy")
loaded_y2 = np.load("y2.npy")

if np.array_equal(loaded_y1, loaded_y2):
    print("loaded_y1==loaded_y2")

if np.array_equal(loaded_y1, y1):
    print("loaded_y1==y1")

print("size y1.npy (in Bytes)=", os.stat("y1.npy").st_size)
print("size y2.npy (in Bytes)=", os.stat("y2.npy").st_size)
```

The output of the above script is:

> 
> loaded_y1==loaded_y2
> loaded_y1==y1
> size y1.npy (in Bytes)= 768
> size y2.npy (in Bytes)= 768

You can see that the size of the two files is equal, and this is what users are expecting from saving a narrow tensor as well.

**Comment by albanD** [comment]
This is linked to sharing views across serialization.
You can see previous issues https://github.com/pytorch/pytorch/issues/40157 and https://github.com/pytorch/pytorch/issues/33101 for example.

You can see full details about what happens in the serialization note of the doc: https://pytorch.org/docs/master/notes/serialization.html#saving-and-loading-tensors-preserves-views
